lib/state_change_syncer.go

```diff
@@ -0,0 +1,726 @@
+		// If not, reset the mempool maps and file, and start from scratch. The consumer will revert the mempool transactions
+		// it currently has and sync from scratch.
+		if stateChangeSyncer.BlockSyncFlushId != event.BlockSyncFlushId {
+			stateChangeSyncer.ResetMempool()
+			return
+		}
+
+		// Check to see if any of the keys in the mempool key set are not in the mempool key value map.
+		// If so, reset the mempool key value map, mempool key set, and mempool file, start from scratch.
+		for key, _ := range stateChangeSyncer.MempoolKeyValueMap {
+			// If any of the keys that the mempool is currently tracking weren't included in the flush, the state syncer
+			// mempool is bad and needs to be reset.
+			if _, ok := stateChangeSyncer.MempoolFlushKeySet[key]; !ok {
+				stateChangeSyncer.ResetMempool()
+				return
+			}
+		}
+		stateChangeSyncer.MempoolFlushKeySet = make(map[string]bool)
+	}
+
+	err := stateChangeSyncer.FlushTransactionsToFile(event)
+	if err != nil {
+		glog.Errorf("StateChangeSyncer._handleStateSyncerFlush: Error flushing transactions to file: %v", err)
+	}
+
+	if !event.IsMempoolFlush {
+		// After flushing blocksync transactions to file, reset the block sync flush ID, and reset the mempool.
+		stateChangeSyncer.BlockSyncFlushId = uuid.New()
+		stateChangeSyncer.ResetMempool()
+	}
+}
+
+func (stateChangeSyncer *StateChangeSyncer) ResetMempool() {
+	stateChangeSyncer.MempoolKeyValueMap = make(map[string][]byte)
+	stateChangeSyncer.MempoolFlushKeySet = make(map[string]bool)
+	delete(stateChangeSyncer.UnflushedBytes, stateChangeSyncer.MempoolFlushId)
+	stateChangeSyncer.MempoolFlushId = uuid.Nil
+	// Truncate the mempool files.
+	stateChangeSyncer.StateChangeMempoolFile.Truncate(0)
+	stateChangeSyncer.StateChangeMempoolIndexFile.Truncate(0)
+	stateChangeSyncer.StateChangeMempoolFileSize = 0
+}
+
+// Add a transaction to the queue of transactions to be flushed to disk upon badger db flush.
+func (stateChangeSyncer *StateChangeSyncer) addTransactionToQueue(flushId uuid.UUID, writeBytes []byte) {
+	stateChangeSyncer.EntryCount++
+
+	unflushedBytes, exists := stateChangeSyncer.UnflushedBytes[flushId]
+	if !exists {
+		unflushedBytes = UnflushedStateSyncerBytes{
+			StateChangeBytes:            []byte{},
+			StateChangeBytesSize:        0,
+			StateChangeOperationIndexes: []uint64{},
+		}
+	}
+
+	unflushedBytes.StateChangeBytes = append(unflushedBytes.StateChangeBytes, writeBytes...)
+
+	// Get the byte index of where this transaction occurs in the unflushed bytes, and add it to the list of
+	// indexes that should be written to the index file.
+	dbOperationIndex := unflushedBytes.StateChangeBytesSize
+	unflushedBytes.StateChangeOperationIndexes = append(unflushedBytes.StateChangeOperationIndexes, dbOperationIndex)
+
+	// Update the state change file size.
+	transactionLen := uint64(len(writeBytes))
+	unflushedBytes.StateChangeBytesSize += transactionLen
+
```

Comment:
I don't understand why we need StateChangeBytesSize. Isn't it always just equal to len(StateChangeBytes)?